File: container_images/hello_world/Dockerfile.py
FROM python:alpine as builder
WORKDIR /hello
#RUN apk add --no-cache binutils patchelf gcc musl-dev
#RUN pip install --no-cache-dir pyinstaller staticx
RUN apk add --no-cache binutils
RUN pip install --no-cache-dir pyinstaller
COPY hello.py .
RUN pyinstaller --onefile hello.py -n hello
CMD ["/hello/dist/hello"]

# PyInstaller builds aren't fully static (libc and ld.so are still required), through staticx that can be fixed
# https://github.com/pyinstaller/pyinstaller/wiki/FAQ#gnulinux

# docker build -t hello:py -f Dockerfile.py .
# docker container run --rm hello:py
# docker images hello:py
# docker image prune --filter label=stage=builder

File: container_images/hello_world/Dockerfile.c
#FROM debian:stable-slim as builder
#LABEL stage=builder
#RUN apt update && apt install -y gcc
#COPY hello.c /
#RUN gcc -o helloworld -static-libgcc -static hello.c
# ~ 762kB image

FROM alpine as builder
LABEL stage=builder
RUN apk add --no-cache gcc musl-dev
COPY hello.c /
RUN gcc -o helloworld -static-libgcc -static hello.c
 
FROM scratch
COPY --from=builder /helloworld /
CMD ["/helloworld"]


# docker build -t hello:c -f Dockerfile.c .
# docker container run --rm hello:c
# docker images hello:c
# docker image prune --filter label=stage=builder
